CertiKProject/slither-certik
# Certik merge 0.9.2

## Notes

This PR merges Slither 0.9.2 into CertiK's Slither repo.

When merging in 0.9.2, I found two conflicts:

### slither/slithir/utils/ssa.py -- the `CallEvent` case of the `copy_ir` function

CertiK modified the EventCall IR instruction to store the full event variable rather than just its name. This allows us to format a better output message in the `unused-event` detector.
(https://github.com/CertiKProject/slither-certik/commit/cdb4e8056fa1af9e74c5b6a9db5c374f42179feb)

Crytic fixed an issue where the event arguments weren't being properly copied over.
(https://github.com/crytic/slither/commit/8ad3ea7555edc9b038c3b098ab3bf616c0f6bb0f)

This PR rewrites the relevant code to include both features.

### using_for handling

CertiK modified the parsing of `using_for` statements to record their positions. It does this by mapping each type `ty` to a list of all (libraryType, loc) pairs corresponding to `using library for ty` statements where `libraryType` is the `UserDefinedType` corresponding to the library and `loc` is the source location of the using statement. Note that `loc` is actually a `Structure` instance; this is a hack as we are only using the `Structure` for its `SourceMapping` components to record the location of the using statement. 
(https://github.com/CertiKProject/slither-certik/commit/6d71e15b34a9f07bb2c510a6f4ab6b2ffc2d7f2e)

Crytic added a few fixes related to using for:
* They added support for `using {functionName1, functionName2, ... functionNameN} for type` statemets. Previously, I think these would just cause Slither log error messages.
* They added support for top-level using statements.

(https://github.com/crytic/slither/commit/f0439905e02b6f6c687492fb0dcdd6a4c4e35ee4)

I have updated the code to include both features. Code for parsing "function list style" using-for statements now records source mapping as well; this feature isn't really necessary because none of the detectors that process using-for statements have any reason to care about "function list style" statements.

One aspect of these conflicting changes still remains unintegrated: the interface for accessing top-level using-for statements does not provide access to their source locations. This is not yet an issue because our detectors do not access top-level using-for statements. Of course, we should extend our detectors to handle top-level using-for statements, but that is outside the scope of this PR. 

## Testing
* Check out `slither-task` and make this a subrepo
* Disable the new detectors named `immutable-states`,`constable-states`, and `var-read-using-this` by adding them to the `disabled_detectors` list in `run-slither.sh`
* Run `./evaluate.sh run 100` in the `tools-exec` and verify that all tests pass

## Related Issue
https://github.com/CertiKProject/slither-task/issues/218

## Additional Comments
This PR breaks the unused contract detector. When we update slither submodule commit of the `slither-task` repo, we will also modify the unused contract detector for compatibility with 0.9.2.
